src/frontend/Ast.ml

```diff
@@ -116,13 +116,32 @@ type 'e transformation =
   | Covariance
 [@@deriving sexp, compare, map, hash]
 
+type ('e, 'm) lhs_with =
+  { assign_identifier: identifier
+  ; assign_indices: 'e index list
+  ; assign_meta: 'm }
+[@@deriving sexp, hash, compare, map]
+
+type 'e untyped_lhs = ('e, located_meta) lhs_with
+[@@deriving sexp, hash, compare, map]
+
+type typed_lhs_meta =
+  { loc: Middle.location_span sexp_opaque [@compare.ignore]
+  ; id_ad_level: Middle.autodifftype
+  ; id_type_: Middle.unsizedtype
+  ; lhs_ad_level: Middle.autodifftype
+  ; lhs_type_: Middle.unsizedtype }
+[@@deriving sexp, compare, map, hash]
+
+type 'e typed_lhs = ('e, typed_lhs_meta) lhs_with
+[@@deriving sexp, hash, compare, map]
+
 (** Statement shapes, where we substitute untyped_expression and untyped_statement
     for 'e and 's respectively to get untyped_statement and typed_expression and
     typed_statement to get typed_statement    *)
-type ('e, 's) statement =
+type ('e, 's, 'l) statement =
   | Assignment of
-      { assign_identifier: identifier
-      ; assign_indices: 'e index list
+      { assign_lhs: 'l
```

Comment:
This would be more natural with higher-kinded polymorphism:
```
assign_lhs: 'e 'l
```
but OCaml doesn't support that.